File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function solution(diffs, times, limit) {
    let timeSum = [times[0]]
    for (let i=1; i<times.length; i++) {
        timeSum.push(times[i]+times[i-1])
    }

    let start = 1; 
    let end = 100000;
    let result  = end
    
    while (start<=end) {
        let level = Math.floor((start+end)/2);
        
        let time = 0;
        for (let i=0; i<diffs.length; i++) {
            if (diffs[i] < level) {
                time+=times[i];
            }
            else {
                time += (diffs[i]-level) * timeSum[i] + times[i];
            }
        }
        
        if (time > limit) {
            start = level + 1
        }
        else {
            end = level - 1
        }
    }
    return start;
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.cpp
#include <string>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

int solution(vector<int> diffs, vector<int> times, long long limit) {
    int left = 1;
    
    int max = *max_element(diffs.begin(), diffs.end());
    int right = max;
    while (left < right) {
        int mid = (left + right) / 2;
        long long limits = limit;
        
        for (int i=0; i<diffs.size(); i++) {
            if (diffs[i]<= mid) limits-=times[i];
            else {
                limits -= times[i];
                if (i>0) {
                    limits -= abs(diffs[i] - mid) * (times[i-1] + times[i]);
                } else {
                    limits -= abs(diffs[i] - mid) * times[i];
                }
            }
            if (limits < 0) break;
        }
        if (limits < 0) left = mid + 1;
        else right = mid;
    }
   
    return left;
}